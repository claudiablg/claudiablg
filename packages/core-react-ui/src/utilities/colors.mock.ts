export const generatedColorCSSVarNames = [
  '--color-transparent',
  '--color-primary-100',
  '--color-primary-300',
  '--color-primary-500',
  '--color-primary-700',
  '--color-primary-900',
  '--color-accent1-100',
  '--color-accent1-300',
  '--color-accent1-500',
  '--color-accent1-700',
  '--color-accent1-900',
  '--color-accent2-100',
  '--color-accent2-300',
  '--color-accent2-500',
  '--color-accent2-700',
  '--color-accent2-900',
  '--color-accent3-100',
  '--color-accent3-300',
  '--color-accent3-500',
  '--color-accent3-700',
  '--color-accent3-900',
  '--color-grey-0',
  '--color-grey-25',
  '--color-grey-50',
  '--color-grey-100',
  '--color-grey-200',
  '--color-grey-300',
  '--color-grey-400',
  '--color-grey-500',
  '--color-grey-600',
  '--color-grey-700',
  '--color-grey-800',
  '--color-grey-900',
  '--color-grey-1000',
  '--color-grey-0-reversed',
  '--color-grey-100-reversed',
  '--color-grey-200-reversed',
  '--color-effect-transparent-light',
  '--color-effect-transparent-medium',
  '--color-effect-transparent-heavy',
  '--color-utility-green-100',
  '--color-utility-green-300',
  '--color-utility-green-500',
  '--color-utility-green-700',
  '--color-utility-green-900',
  '--color-utility-yellow-100',
  '--color-utility-yellow-300',
  '--color-utility-yellow-500',
  '--color-utility-yellow-700',
  '--color-utility-yellow-900',
  '--color-utility-red-100',
  '--color-utility-red-300',
  '--color-utility-red-500',
  '--color-utility-red-700',
  '--color-utility-red-900',
  '--color-intent-accessibility-color',
  '--color-intent-transparent',
  '--color-intent-primary',
  '--color-intent-primary-reversed',
  '--color-intent-secondary',
  '--color-intent-secondary-reversed',
  '--color-intent-primary-text',
  '--color-intent-primary-text-reversed',
  '--color-intent-secondary-text',
  '--color-intent-secondary-text-reversed',
  '--color-intent-tertiary-text',
  '--color-intent-tertiary-text-reversed',
  '--color-intent-success-text',
  '--color-intent-success-action',
  '--color-intent-success-background',
  '--color-intent-warning-text',
  '--color-intent-warning-action',
  '--color-intent-warning-background',
  '--color-intent-danger-text',
  '--color-intent-danger-action',
  '--color-intent-danger-background',
  '--color-intent-background0',
  '--color-intent-background1',
  '--color-intent-background2',
];
