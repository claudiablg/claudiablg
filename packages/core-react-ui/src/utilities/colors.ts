import { Color } from '@newrade/core-design-system';

export function getHSLColor({ h, s, l, a }: Partial<Color> = { h: 200, s: 100, l: 100, a: 1 }) {
  return `hsla(${h}, ${s}%, ${l}%, ${a || 1})`;
}

export function lightenColor(color: Color, amount: number) {}

export function darkenColor(color: Color, amount: number) {}
