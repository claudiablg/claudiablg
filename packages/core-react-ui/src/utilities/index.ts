export * from './colors.utilities';
export * from './component.utilities';
export * from './css-variable.utilities';
export * from './effects.utilities';
export * from './filter-color.utility';
export * from './text.utilities';
export * from './theme.utilities';
export * from './utilities';
