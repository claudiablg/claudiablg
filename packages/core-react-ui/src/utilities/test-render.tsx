import * as React from 'react';
import { render } from '@testing-library/react';
import { TreatProvider } from 'react-treat/lib/types';

export function renderTest(Component: any) {
  return render(<TreatProvider theme={'light'}>{Component}</TreatProvider>);
}
