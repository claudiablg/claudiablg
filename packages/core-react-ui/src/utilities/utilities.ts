import { ColorIntents, Colors } from '@newrade/core-design-system';

/**
 * Create default color intents.
 */
export function createDefaultColorIntents(colors: Colors): ColorIntents {
  return {
    primary: colors.primary500,
    primaryReversed: colors.greyscale0,
    secondary: colors.primary500,
    secondaryReversed: colors.greyscale0,
    primaryText: colors.greyscale900,
    primaryTextReversed: colors.greyscale0,
    secondaryText: colors.greyscale700,
    secondaryTextReversed: colors.greyscale0,
    tertiaryText: colors.greyscale600,
    tertiaryTextReversed: colors.greyscale0,
    successText: colors.utilityGreen1000,
    successAction: colors.utilityGreen500,
    successBackground: colors.utilityGreen100,
    warningText: colors.utilityYellow1000,
    warningAction: colors.utilityYellow500,
    warningBackground: colors.utilityYellow100,
    dangerText: colors.utilityRed1000,
    dangerAction: colors.utilityRed500,
    dangerBackground: colors.utilityRed100,
    background0: colors.greyscale0,
    background1: colors.greyscale100,
    background2: colors.greyscale200,
  };
}

/**
 * Convert a number (in px) into a rem value.
 * @param px value in pixel
 * @param baseUnitPx font-size set on the <html/> element
 */
export function pxToRem({ baseUnitPx, value }: { baseUnitPx: number; value: number }): string {
  return `${value / baseUnitPx}rem`;
}

/**
 * Convert a number (in px) into a em relative value.
 * @param px value in pixel
 * @param fontSize font-size set on the element
 */
export function pxToEm({ fontSize, value }: { fontSize: number; value: number }): string {
  return `${value / fontSize}em`;
}

/**
 * Convert px value (e.g. '15px') into a rem value.
 * @param px value in string px value
 * @param baseUnitPx font-size set on the <html/> element
 */
export function pxStringToRem({ baseFontSize, value }: { baseFontSize: number; value: string }): string {
  if (!value?.length) {
    return '';
  }

  const pxPattern = /(\d+\.?(\d+))/g;
  const match = pxPattern.exec(value);
  if (!(match?.length && match[0]?.length)) {
    return '';
  }
  const conversion = Number(match[0]);
  if (!conversion) {
    return '';
  }

  return `${conversion / baseFontSize}rem`;
}

/**
 * Convert px value (e.g. '15px') into a number value.
 * @param px value in string px value
 */
export function pxStringToNumber({ value }: { value: string }): number | undefined {
  if (!value?.length) {
    return undefined;
  }

  const pxPattern = /(\d+\.?(\d+))/g;
  const match = pxPattern.exec(value);
  if (!(match?.length && match[0]?.length)) {
    return undefined;
  }
  const conversion = Number(match[0]);
  if (conversion === undefined || conversion === NaN || conversion === null) {
    return undefined;
  }

  return conversion;
}
