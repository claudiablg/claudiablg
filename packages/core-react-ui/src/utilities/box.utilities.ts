import { Border, BorderStyles, BoxStyle, Radius } from '@newrade/core-design-system';
import { getCSSColor } from './colors.utilities';
import { keys, px } from './utilities';

export function getCSSBoxStyle(box: BoxStyle): BoxStyle<string> {
  return {
    width: px({ value: box.width }),
    height: px({ value: box.height }),
    // padding: box.padding,
    border: getCSSBorderStyles(box.border),
    // outline: box.outline,
    // backgroundColor: box.backgroundColor,
  };
}

export function getCSSBorderStyles(borders?: BorderStyles): BorderStyles<string> | undefined {
  if (!borders) {
    return undefined;
  }

  return keys(borders).reduce((previous, current) => {
    return { ...previous, [current]: getCSSBoxBorder(borders[current]) };
  }, {} as BorderStyles<string>);
}

export function getCSSBoxBorder(border?: Border): Border<string> {
  if (!border) {
    return {};
  }

  return {
    color: border.color ? getCSSColor(border.color) : undefined,
    width: border.width ? px({ value: border.width }) : undefined,
    style: border.style || 'solid',
    radius: getCSSBoxRadius(border.radius),
  };
}

/**
 * @see https://developer.mozilla.org/en-US/docs/Web/CSS/border-radius
 */
export function getCSSBoxRadius(radius?: Radius): string {
  if (!radius) {
    return '';
  }

  return radius.all
    ? `${radius.all}`
    : `${radius.topLeft} ${radius.topRight} ${radius.bottomRight} ${radius.bottomLeft}`;
}
