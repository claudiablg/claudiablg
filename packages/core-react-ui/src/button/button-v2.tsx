import {
  ButtonIcon,
  ButtonProps,
  ButtonSize,
  LABEL_SIZE,
  TEXT_STYLE,
} from '@newrade/core-design-system';
import React, { ButtonHTMLAttributes, useRef } from 'react';
import { usePreventPinchZoom } from '../hooks/use-prevent-pinch-zoom';
import { CommonComponentProps } from '../props/component-common.props';
import { Label } from '../text/label';
import { getDefaultTextFromProps, getMergedClassname } from '../utilities/component.utilities';
import * as styles from './button-v2.css';

type Props = CommonComponentProps &
  ButtonHTMLAttributes<any> &
  Pick<ButtonProps, 'icon' | 'role' | 'size' | 'state' | 'variant'> & {
    as?: 'button' | 'a' | 'div';
  } & {
    disabled?: boolean;
    loading?: boolean;
    Icon?: React.ReactNode;
    dataPressed?: boolean;
    collapsePadding?: 'left' | 'right';
  };

/**
 * General purpose button element, can be rendered using the AsElement={} prop.
 * @version 2
 */
export const ButtonV2 = React.forwardRef<any, Props>(
  (
    {
      id,
      style,
      className,
      children,
      variant,
      collapsePadding,
      as,
      AsElement,
      size,
      state,
      disabled,
      icon = ButtonIcon.right,
      Icon,
      ...props
    },
    forwardedRef
  ) => {
    const localRef = useRef<HTMLButtonElement>(null);
    const ref = forwardedRef ? (forwardedRef as React.RefObject<HTMLButtonElement>) : localRef;
    const type = as ? as : 'button';

    /**
     *
     * Icon
     *
     */

    const iconClassNames = getMergedClassname([
      // styles.base,
      // icon === ButtonIcon.right ? styles.right : '',
      // icon === ButtonIcon.left ? styles.left : '',
      // icon === ButtonIcon.icon ? styles.icon : '',
    ]);

    const IconSvg =
      icon && icon !== ButtonIcon.none && Icon
        ? React.cloneElement(Icon as React.ReactElement, {
            className: iconClassNames,
            preserveAspectRatio: `xMinYMin meet`,
          })
        : null;

    /**
     * Event handling
     */
    usePreventPinchZoom(ref.current);

    /**
     * Styles
     */
    const classNames = getMergedClassname([
      className,
      styles.base,
      styles[size ? size : ButtonSize.medium],
      // styles.primary,
      // styles[variant ? variant : Variant.primary],
    ]);
    const renderedChildren = children
      ? children
      : getDefaultTextFromProps('button', {
          variant,
          size,
          icon,
          disabled,
        });

    function getLabelSizeForButtonSize(size?: ButtonSize): LABEL_SIZE {
      switch (size) {
        case ButtonSize.large: {
          return LABEL_SIZE.medium;
        }
        case ButtonSize.medium: {
          return LABEL_SIZE.small;
        }
        default:
        case ButtonSize.small: {
          return LABEL_SIZE.small;
        }
        case ButtonSize.xSmall: {
          return LABEL_SIZE.xSmall;
        }
      }
    }

    const CustomElement = AsElement
      ? React.cloneElement(
          AsElement as React.ReactElement,
          {
            id,
            style,
            className: classNames,
            ref: ref,
            dataicon: `${icon}`,
            datapaddingcollapse: `${collapsePadding}`,
            ...props,
          },
          <>
            {icon === ButtonIcon.icon ? null : (
              <Label
                variantDisplay={'inline'}
                variantStyle={TEXT_STYLE.bold}
                variant={getLabelSizeForButtonSize(size)}
              >
                {renderedChildren}
              </Label>
            )}
            {IconSvg}
          </>
        )
      : null;

    if (CustomElement) {
      return CustomElement;
    }

    return (
      <button
        id={id}
        style={style}
        className={classNames}
        ref={ref}
        disabled={disabled}
        // @ts-ignore
        dataicon={`${IconSvg ? icon : ''}`}
        datapaddingcollapse={`${collapsePadding}`}
        {...props}
      >
        {icon === ButtonIcon.icon ? null : (
          <Label
            variantDisplay={'inline'}
            variantStyle={TEXT_STYLE.bold}
            variant={getLabelSizeForButtonSize(size)}
          >
            {renderedChildren}
          </Label>
        )}
        {IconSvg}
      </button>
    );
  }
);
