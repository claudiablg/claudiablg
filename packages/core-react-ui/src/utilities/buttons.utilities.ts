import * as DS from '@newrade/core-design-system';
import { ButtonColors, ButtonSizes, ButtonVariants, DesignSystem } from '@newrade/core-design-system';
import { kebab } from 'case';
import { CSSButtons, CSSButtonsVarNames } from '../design-system/css-buttons';
import { getCSSBoxStyle } from './box.utilities';
import { getCSSColor, getCSSVarForColors, getCSSVarNamesForColors } from './colors.utilities';
import { keys } from './utilities';

export function getCSSButtons(theme: DesignSystem): CSSButtons {
  const { colors, components } = theme;
  const { buttons } = components;

  const colorVar = getCSSVarForColors({
    colors: colors.colors,
    colorIntents: colors.colorIntents,
  });
  const colorVarNames = getCSSVarNamesForColors({
    colors: colors.colors,
    colorIntents: colors.colorIntents,
  });

  const buttonsVar = getCSSVarNameForButtons({
    buttons,
    prefix: 'btn',
    varBrackets: true,
  });
  const buttonsVarNames = getCSSVarNameForButtons({
    buttons,
    prefix: 'btn',
    varBrackets: false,
  });

  return {
    var: buttonsVar,
    varNames: buttonsVarNames,
    variants: getCSSButtonVariants(buttons.variants),
    sizes: getCSSButtonSizes(buttons.sizes),
  };
}

/**
 * Creates css variables from Buttons properties
 */
function getCSSVarNameForButtons({
  buttons,
  prefix,
  varBrackets,
}: {
  buttons: DS.Buttons;
  prefix: string;
  varBrackets: boolean;
}): CSSButtonsVarNames {
  const buttonsVarNames = {
    sizes: {},
    variants: {},
  } as CSSButtonsVarNames;
  const { variants, sizes } = buttons;

  keys(variants).forEach((variant) => {
    const currentVariant = variant as keyof CSSButtons['variants'];
    const formattedCurrentVariant = kebab(currentVariant)
      .replace('primary', 'prim')
      .replace('secondary', 'sec')
      .replace('tertiary', 'ter');
    if (currentVariant && currentVariant.length) {
      // handle variant
      if (variants[currentVariant] && typeof variants[currentVariant] === 'object') {
        const buttonColors = variants[currentVariant] as ButtonColors;
        keys(buttonColors).forEach((prop) => {
          const formattedPropName = kebab(prop);
          // @ts-ignore
          if (!buttonsVarNames['variants'][variant]) {
            // @ts-ignore
            buttonsVarNames['variants'][variant] = {};
          }
          if (varBrackets) {
            // @ts-ignore
            buttonsVarNames['variants'][variant][
              prop
            ] = `var(--${prefix}-${formattedCurrentVariant}-${formattedPropName})`;
            return;
          }

          // @ts-ignore
          buttonsVarNames['variants'][variant][prop] = `--${prefix}-${formattedCurrentVariant}-${formattedPropName}`;
          return;
        });
      }
    }
  });

  keys(sizes).forEach((variant) => {
    const currentVariant = variant as keyof CSSButtons['variants'];
    const formattedCurrentVariant = kebab(currentVariant)
      .replace('primary', 'prim')
      .replace('secondary', 'sec')
      .replace('tertiary', 'ter');
    if (currentVariant && currentVariant.length) {
      // handle variant
      if (variants[currentVariant] && typeof variants[currentVariant] === 'object') {
        const buttonColors = variants[currentVariant] as ButtonColors;
        keys(buttonColors).forEach((prop) => {
          const formattedPropName = kebab(prop);
          // @ts-ignore
          if (!buttonsVarNames['sizes'][variant]) {
            // @ts-ignore
            buttonsVarNames['sizes'][variant] = {};
          }
          if (varBrackets) {
            // @ts-ignore
            buttonsVarNames['sizes'][variant][
              prop
            ] = `var(--${prefix}-${formattedCurrentVariant}-${formattedPropName})`;
            return;
          }

          // @ts-ignore
          buttonsVarNames['sizes'][variant][prop] = `--${prefix}-${formattedCurrentVariant}-${formattedPropName}`;
          return;
        });
      }
    }
  });

  return buttonsVarNames;
}

function getCSSButtonVariants(buttons: ButtonVariants): ButtonVariants<string> {
  return keys(buttons).reduce((previous, current) => {
    const variantColorStyles: ButtonColors<string> = {
      backgroundColor: getCSSColor(buttons[current].backgroundColor),
    };
    previous = {
      ...previous,
      [current]: variantColorStyles,
    };
    return previous;
  }, {} as ButtonVariants<string>);
}

function getCSSButtonSizes(buttons: ButtonSizes): ButtonSizes<string> {
  return keys(buttons).reduce((previous, current) => {
    previous = {
      ...previous,
      [current]: getCSSBoxStyle(buttons[current]),
    };
    return previous;
  }, {} as ButtonSizes<string>);
}
