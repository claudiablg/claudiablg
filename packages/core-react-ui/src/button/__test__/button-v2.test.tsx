import React from 'react';
import { renderTest } from '../../utilities/test-render';
import { ButtonV2 } from '../button-v2';

describe('the button component (v2)', () => {
  it('should render correctly', () => {
    const { container } = renderTest(<ButtonV2>Hello world</ButtonV2>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
