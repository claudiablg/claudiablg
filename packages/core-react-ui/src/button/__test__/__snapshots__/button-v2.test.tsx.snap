// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the button component (v2) should render correctly 1`] = `
<button
  class="button-v2__1ypu3sr0 button-v2__1ypu3sr5"
  dataicon=""
  datapaddingcollapse="undefined"
>
  <div
    class="style_theme style_theme style_theme style_theme"
  >
    Hello world
  </div>
</button>
`;
